File: src/world.dart
import 'dart:html';
import 'dart:math';

import 'enemies.dart';
import 'entity.dart';

class World {
  num width, height;
  List<Entity> _entities = [], _particles = [];

  World({this.width = 0, this.height = 0});

  Iterable<Entity> getEnemies() => _entities.where((e) => e is Enemy);

  Iterable<Entity> getEntitiesNear(Entity entity) {
    // TODO: only return actors close to given actor
    // Only return entities we can collide with
    return _entities.where((e) => e != entity && e.isAlive && e.damage > 0);
  }

  void addEntity(Entity entity) => entity.damage == 0 ? _particles.add(entity) : _entities.add(entity);

  void spawnInBounds(Entity entity) {
    final location = getEmptySpawnLocation(entity.radius);
    entity.spawn(x: location.x, y: location.y);
    addEntity(entity);
  }

  void spawnOutOfBounds(Entity entity) {
    final location = getLocationOutside();
    entity.spawn(x: location.x, y: location.y);
    addEntity(entity);
  }

  Point getEmptySpawnLocation(num radius) {
    final random = new Random();

    const MAX_ATTEMPTS = 10;
    for (var i = 0; i < MAX_ATTEMPTS; i ++) {
      final point = new Point(random.nextDouble() * width, random.nextDouble() * height);
      final overlapping = _entities.where((e) => e.distanceFromPoint(point.x, point.y) < e.radius + radius);
      if (overlapping.length == 0) {
        return point;
      }
    }

    // If all else fails, put it outside. TODO: something better?
    return getLocationOutside();
  }

  Point getLocationOutside() {
    final angle = new Random().nextDouble() * pi * 2;
    return new Point(width/2 + cos(angle) * width/2 * 1.5, height/2 + sin(angle) * height/2 * 1.5);
  }

  void update(num dt) {
    // Use local list of entities in case more are added during update
    final localEntities = List<Entity>.from(_entities);
    localEntities.forEach((e) => e.update(dt, this));
    
    // Perform each entity vs entity check only once
    final others = List<Entity>.from(localEntities);

    for (var e in localEntities) {
      others.remove(e);

      // TODO: What if we hit multiple others? Figure out which we hit first?
      for (var o in others) {
        final hitTime = e.timeUntilHit(o);

        if (-dt < hitTime && hitTime < 0) {
          e.updatePosition(hitTime);
          o.updatePosition(hitTime);

          Entity.handleBounce(e, o);
          e.hitWith(o, this);
          o.hitWith(e, this);

          e.updatePosition(-hitTime);
          o.updatePosition(-hitTime);
        }
      }
    }
    _entities.removeWhere((e) => !e.isAlive);

    _particles..forEach((p) => p.update(dt, this))..removeWhere((p) => !p.isAlive);
  }

  void draw(CanvasRenderingContext2D ctx) {
    _entities.forEach((e) => e.draw(ctx));
    _particles.forEach((p) => p.draw(ctx));
  }
}

File: src/world.js
import { Ship } from './Ship.js';
import { Rock } from './Rock.js';
import { Info } from '../info/info.js';
import * as Util from './Util.js';

const SPAWN_GAP = 2.5;

export class World {
  size;
  entities = [];
  particles = [];

  level;
  player;

  #levelTime = 0;
  #waveIndex = 0;

  constructor( level ) {
    this.level = level;
    this.size = level?.size ?? 500;
  }

  spawnInside( entity ) {
    [ entity.x, entity.y ] = this.getEmptyLocation( entity );
    this.entities.push( entity );
  }

  spawnOutside( entity, angle ) {
    [ entity.x, entity.y ] = this.getEmptyLocation( entity, this.size * 1.0, angle );
    this.entities.push( entity );
  }

  getEmptyLocation( entity, radius = Math.random() * this.size, angle = Math.random() * Math.PI * 2, spread = Math.PI / 3) {
    let x, y, tries = 0;
    do {
      const ang = angle + spread * ( Math.random() - 0.5 );
      x = Math.cos( ang ) * radius;
      y = Math.sin( ang ) * radius;
      tries ++;
    }
    while( tries < 20 && this.entities.some( e => Math.hypot( e.x - x, e.y - y ) < entity.size * SPAWN_GAP ) );

    // if ( tries == 20 ) {
    //   debugger;
    // }

    return [ x, y ];
  }

  #updateLevel( dt ) {
    this.#levelTime += dt;

    if ( this.#waveIndex < this.level.waves.length && this.level.waves[ this.#waveIndex ].time <= this.#levelTime ) {
      this.level.waves[ this.#waveIndex ].spawns.forEach( spawn => {
        const ship = new Ship( Info[ spawn.type ] );
        [ ship.x, ship.y ] = Util.rotatedXY( this.level.size + spawn.x, spawn.y, this.level.entryAngle );
        
        // Head straight across the map
        [ ship.goalX, ship.goalY ] = Util.rotatedXY( this.level.size + spawn.x, -spawn.y, this.level.exitAngle );
        ship.angle = Math.atan2( ship.goalY - ship.y, ship.goalX - ship.x );

        this.entities.push( ship );
      } );

      this.#waveIndex ++;
    }
  }

  #handleRocks() {
    // "Bounce" out-of-bounds rocks (to try to keep same number of rocks in level)
    // TODO: Should we add rocks when other rocks are destoryed? Or just randomly?
    this.entities.filter( e => e instanceof Rock).forEach( rock => {
      if ( rock.x < -this.size - rock.size )  rock.dx = -rock.dx;
      if ( rock.y < -this.size - rock.size )  rock.dy = -rock.dy;
      if ( this.size + rock.size < rock.x )   rock.dx = -rock.dx;
      if ( this.size + rock.size < rock.y )   rock.dy = -rock.dy;
    } );
  }

  update( dt ) {
    if ( this.level ) { 
      this.#updateLevel( dt );
    }

    
    this.entities.forEach( entity => entity.update( dt ) );
    
    this.#handleRocks();
    this.entities.filter( e => e instanceof Ship).forEach( enemy => enemy.think( this.player, this ) );
    
    // TODO: Not everything checks against everything else...do these by category?
    
    const createdEntities = [], createdParticles = [];
    let others = this.entities;
    this.entities.forEach( entity => {
      others = others.filter( other => other != entity );

      others.forEach( other => {
        if ( entity != other.owner && other != entity.owner && entity.distanceTo( other ) < 0 ) {
          // TODO: Find actual point of impact, and normal
          const norm = Math.atan2( other.y - entity.y, other.x - entity.x );
          const hit = { 
            x: other.x - Math.cos( norm ) * other.size,
            y: other.y - Math.sin( norm ) * other.size, 
            normal: norm,
          };
          entity.hitWith( hit, other );
          other.hitWith( hit, entity );

          // No friction or elasticity in this game
          const normX = Math.cos( norm ), normY = Math.sin( norm );
          const p = 2 * ( ( entity.dx - other.dx ) * normX + 
                          ( entity.dy - other.dy ) * normY ) / 
                        ( entity.mass + other.mass );
          
          entity.dx -= p * other.mass * normX;
          entity.dy -= p * other.mass * normY;
          other.dx += p * entity.mass * normX;
          other.dy += p * entity.mass * normY;
        }
      } );
      
      createdEntities.push( ...entity.createdEntities.splice( 0 ) );
      createdParticles.push( ...entity.createdParticles.splice( 0 ) );
    } );
    
    this.particles.forEach( part => part.update( dt ) );
    
    this.entities.push( ...createdEntities );
    this.particles.push( ...createdParticles );
    
    this.entities = this.entities.filter( e => e.isAlive() );
    this.particles = this.particles.filter( p => p.isAlive() );
  }

  draw( ctx ) {
    // DEBUG
    // ctx.beginPath();
    // ctx.moveTo( -this.size, 0 );  ctx.lineTo( this.size, 0 );
    // ctx.moveTo( 0, -this.size );  ctx.lineTo( 0, this.size );
    // for ( let rad = 0; rad <= this.size; rad += 100 ) {
    //   ctx.moveTo( rad, 0 );
    //   ctx.arc( 0, 0, rad, 0, Math.PI * 2 );
    // }
    // ctx.strokeStyle = 'gray';
    // ctx.stroke();

    this.particles.forEach( p => p.draw( ctx ) );
    this.entities.forEach( e => e.draw( ctx ) );
  }

  drawMinimap( ctx ) {
    ctx.clearRect( 0, 0, ctx.canvas.width, ctx.canvas.height );

    // Assume a square minimap
    const scale = ctx.canvas.width / ( this.size * 2 );
    
    this.entities.filter( e => e.size > 8 ).forEach( e => {
      ctx.beginPath();
      ctx.arc( 
        ( this.size + e.x ) * scale, 
        ( this.size + e.y ) * scale, 
        Math.max( 1, e.size * scale ), 
        0, Math.PI * 2 );
      ctx.fillStyle = e.bodyFill;
      ctx.fill();
    } );
  }
}

